Feature/UI memory

gbalan-microej opened the pull request.

rlouedec-microej commented on VEEPortingGuide/vgImage.rst:
```diff
@@ -73,22 +73,56 @@ Resource Vector Image
-However, the RAW image data must be copied into a byte-addressable memory to use it. 
-No data is stored in the Java heap: the image data should be copied into the MicroUI image heap.
+When the image is stored in an external byte-addressable memory, the image cycle-life is identical as the compile-time image.
+However, when the image is stored the in an external memory which is not byte-addressable, the RAW image data must be copied into a byte-addressable memory to use it. 
+No data is stored in the Java heap: the image data should be copied into the MicroUI :ref:`images_heap`.
 The implementation is responsible for the image's lifecycle: allocation and release (already implemented in the :ref:`section_vg_cco`).
```
```suggestion
For an image loaded from an external memory which is not byte-addressable, its data must be copied into byte-addressable memory before the image can be used for drawings.
By default (see :ref:`section_vg_cco`), the image data is copied into MicroUI :ref:`images_heap`.
```

gbalan-microej commented on VEEPortingGuide/uiFontCore.rst:
```diff
@@ -1,57 +1,16 @@
-   }
+The Font Loader module is an on-board engine that 
+
+* retrieves font data that is ready to be displayed without needing additional runtime memory, 
+* retrieves font in external memories (:ref:`section_font_loader_memory` loader).
```
The description has the same look as the image loader (I try to always use the same type of description between the font, image, vg font and vg image modules). For the moment, I'm leaving the commentary as it is, but it may be necessary to reword it for all the chapters.

rlouedec-microej replied:
I noticed that, I did comment only for the font section but this should be applied to others as well 👍

rlouedec-microej replied:
Thanks for your clarification already.
I suggest further rewording: (clarification of the "and/or" + of the packaging of font data)
```diff
- The Font Loader is a module of the MicroUI runtime that:
- 
- * retrieves font data (precomputed bitmaps of glyphs generated by the :ref:`section_fontgen`) that is ready to be displayed without needing additional runtime memory, 
- * retrieves font data in external memories (:ref:`section_font_loader_memory` loader).
+ The Font Loader is a module of the MicroUI runtime that loads font data (precomputed bitmaps of glyphs) ready to be displayed.
+ The font data must be stored as a resource (in EJF raw format).
+ Typically, these resources are generated by the :ref:`section_fontgen` and embedded as internal resources or loaded from external memories (:ref:`section_font_loader_memory` loader).
```

gbalan-microej replied:
ok (changes accepted) but the intro doesn't look the same as the image loader (https://docs.microej.com/en/latest/VEEPortingGuide/uiImageLoader.html). In fact, it's not a problem, but as I said before, the idea was to use the same convention.

rlouedec-microej replied:
If you want, we can update it accordingly. Here is a suggestion:
```diff
+ The Image Loader is a module of the MicroUI runtime that loads image data (image size, pixel format & bitmap) ready to be displayed.
+ During the image drawing, the Image Renderer will, if needed, convert the bitmap to target pixel format.
+ The image data must be stored as a resource (in MEJ_IMG raw format, e.g. ``*.png_raw``).
+ For byte-addressable image resources, better performances can be achieved if it is memory aligned (according to the bpp).
+ Typically, these resources are generated by the :ref:`section_image_generator` and embedded as internal resources (in which case it is automatically memory aligned) or loaded from external memories (:ref:`section_image_external_memory` loader).
+ It can also dynamically convert image files (such as PNG) into supported format (see `Image Decoder<image_runtime_decoder>`_) and manage MicroUI :ref:`section_buffered_image` created by the Application.
```

gbalan-microej replied:
I prefer the bulleted list (list of characteristics, there's a notion of sequence) even if it's not the same as the font loader (because there aren't many bullets actually). What's more, I only want to talk to the image loader, not the image renderer (and the renderer doesn't convert images, the loader does).

gbalan-microej replied:
sorry my mistake: there **no** notion of sequence ! 

this chapter is dedicated to the image loader which is a part of the image enigne. The functional description of the image engine (generator, loader, renderer) is described in the first chapter "overview" https://docs.microej.com/en/latest/VEEPortingGuide/uiImageOverview.html#functional-description

- the generator generates the images
- the loader loads them (manage external resources, call the runtime decoders if it is required, convert to another format on demande)
- the renderer renders

rlouedec-microej replied:
Oh... This diagram is really nice but I have so many questions now...

To keep it short, it presents a 5 steps process: "generator" (although, btw, I am not a fan of the terminology, I would prefer "pre-processor"), "loader", "decoder", "renderer", "drawing".
Whereas the documentation merge loader/decoder and renderer/drawing.

That may be fine but I have no clue what the "renderer" is doing now.
Probably this is because the diagram is about **resource** image format, not **pixel** format.

I suggest we keep this discussion open and address it later.

gbalan-microej replied:
done, see https://youtrack.cross/issue/M0092MEJAUI-2248/doc-UI3-ideas-of-chapters (+ a copy of the discussion in RC)

gbalan-microej commented on VEEPortingGuide/uiImageOverview.rst:
```diff
@@ -165,6 +165,13 @@ Process overview:
+
+-  MicroUI module (see :ref:`section_microui`),
+-  Display module (see :ref:`section_display`): This module gives
+   the characteristics of the Graphical Engine that are useful to configure the :ref:`section_image_generator`.
```
Originally, you could choose which modules to install in a VEE Port using the .platform file. In this way, each module described its deps to make sure you got a coherent VEE Port. All the internal modules of the Image module (same for the font module: generator, loader and renderer) had their own "dependencies" chapter like all the other modules (VM etc.).  I've simply moved these chapters into a single chapter.

rlouedec-microej replied:
I understand, I am just taking the opportunity to comment it 😃

rlouedec-microej replied:
With my understanding, I can suggest:
```diff
- -  Display module (see :ref:`section_display`): This module gives
-    the characteristics of the Graphical Engine that are useful to configure the :ref:`section_image_generator`.
+ -  Display module (see :ref:`section_display`): the characteristics of the target display are used to configure the :ref:`section_image_generator`.
```

gbalan-microej commented on VEEPortingGuide/vgFont.rst:
```diff
@@ -61,11 +61,49 @@ There are two separate Abstraction Layer API header files (see :ref:`LLVG-FONT-A
+Principle
+~~~~~~~~~
+
+MicroVG does not provide some API to make the distinction between a font loaded from different kind of memories (internal or external, byte-addressable or not).
+The implementation must adjust itself how to load an read a font. The :ref:`CCO MicroVG and Freetype<section_vg_cco>` feature an option to enable the font management from an external memory which is not byte-addressable.
+This option requires an implementation of the :ref:`section_externalresourceloader`.
```
Once again, this chapter exists to have a similarity between the chapters image, the font, the vg image and the vg font. I've reworded it.

gbalan-microej commented on VEEPortingGuide/vgFont.rst:
```diff
@@ -61,11 +61,49 @@ There are two separate Abstraction Layer API header files (see :ref:`LLVG-FONT-A
+9. Freetype (:ref:`section_vg_cco`) recognizes this resource as external resource; it configures itself to manage this resource differently than an internal resource.
+10. The application can use the font.
+
+.. note:: The simulator (Front Panel) does not manage the external resources. All images listed in ``*.externresources.list`` files are copied in the external resources folder, and this folder is added to the simulator's classpath. 
```
hmm you're right. I have moved this note a dedicated chapter "simulation" (idem for image, font, vg image, vg font)

gbalan-microej commented on VEEPortingGuide/vgFont.rst:
```diff
@@ -61,11 +61,49 @@ There are two separate Abstraction Layer API header files (see :ref:`LLVG-FONT-A
+5. Deploy the external resources in the external memory (SDCard, flash, etc.).
+6. (optional) Update the implementation of the :ref:`section_externalresourceloader`.
+7. Build and link the application with the BSP.
+8. The application loads the external resource using `ej.microvg.VectorFont.loadFont()`_.
+9. Freetype (:ref:`section_vg_cco`) recognizes this resource as external resource; it configures itself to manage this resource differently than an internal resource.
+10. The application can use the font.
```
i have added a link to your new section "Principle" into the CCO Freetype chapter

rlouedec-microej replied:
Thanks !

gbalan-microej commented on VEEPortingGuide/vgFont.rst:
```diff
@@ -61,11 +61,49 @@ There are two separate Abstraction Layer API header files (see :ref:`LLVG-FONT-A
+Process
+~~~~~~~
+
+The following steps describe how to open an external resource from the application:
```
maybe, but it's always the same thing: the user wants to read something dedicated to his use case (how to load an external vg font) and making links to a generic section is sometimes disturbing. On the other hand, the microej guy wants generic chapters to simplify the documentation...

rlouedec-microej replied:
Yeah, I know. There is no good answer..
In my personal opinion, it was generic enough to be moved to another section.
Maybe we could keep a dedicated guide that say:
* "follow the steps in the other section"
* <insert the 3 lasts bullets after>

WDYT?

gbalan-microej replied:
yes and no. Yes because a chapter in this page https://docs.microej.com/en/latest/VEEPortingGuide/externalResourceLoader.html must exist and no because there are some differences specific to the module (image, font, or generic resources): name of the external list file, format etc.

gbalan-microej commented on VEEPortingGuide/vgImage.rst:
```diff
@@ -73,22 +73,56 @@ Resource Vector Image
+Filtered Image
+--------------
+
+MicroVG provides the API `VectorImage.filterImage()`_ to decline an image in another image using a 4x5 color matrix.
+The resulting image is a copy of the original (plus color transformation) stored in the MicroUI :ref:`images_heap`.
+The implementation is responsible for the image's lifecycle: allocation and release (already implemented in the :ref:`section_vg_cco`).
```
only available for MicroVG Over VGLite; doc updated

rlouedec-microej replied:
Perfect !

gbalan-microej commented:
@rlouedec-microej I've modified the doc according to your suggestions. Could you take a look at it? I've also corrected something in your new "Principle" chapter (same branch, last commit)

rlouedec-microej commented on ApplicationDeveloperGuide/UI/MicroUI/fonts.rst:
```diff
@@ -491,6 +491,15 @@ Examples:
+--------------
+
+A font can be stored in an :ref:`external memory<chapter.microej.applicationResources>`. 
+The management of this kind of font may be different than the internal fonts and require a dedicated heap.
```
Follow up conversation: https://github.com/MicroEJ/docs/commit/b99ccd1182b51f4c7113a32e57aafcc86f1e4b90#r126534929

rlouedec-microej commented on ApplicationDeveloperGuide/UI/MicroUI/images.rst:
```diff
@@ -475,6 +485,8 @@ The size of the images heap can be configured with the ``ej.microui.memory.image
 
 .. warning:: A `ResourceImage`_  allocated on the images heap must be closed manually by the application (`ResourceImage.close()`_); otherwise, a memory leak will occur on the images heap.
 
+For more detail about the images heap implementation, refers to :ref:`this chapter<section_image_loader_memory>` in the VEE Port Guide. 
```
Typo:
```suggestion
For more details about the images heap implementation, refers to :ref:`this chapter<section_image_loader_memory>` in the VEE Port Guide. 
```

rlouedec-microej commented on ApplicationDeveloperGuide/UI/MicroVG/vectorfont.rst:
```diff
@@ -25,12 +25,13 @@ It also provides metrics measurement methods to correctly place the text within
-
-.. _VectorFont: https://repository.microej.com/javadoc/microej_5.x/apis/
-.. _ej.microvg.VectorFont.loadFont(): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microvg/VectorFont.html#loadFont-java.lang.String-
+The fonts are decoded at compile-time.
+They don't need to be processed by a generator tool.
+Font files must be declared as resources in a ``.resources.list`` file available in the classpath (:ref:`chapter.microej.applicationResources`). 
+To declare them as :ref:`external resources<vectorfont_external>`, the font files must be declared too in a ``.externresources.list`` file.
```
This is wrong, right ?
```suggestion
The fonts are decoded at runtime.
They don't need to be pre-processed by some generator tool like MicroUI Fonts.
Vector Font files must be declared as resources in a ``.resources.list`` file available in the classpath (:ref:`chapter.microej.applicationResources`). 
To declare them as :ref:`external resources<vectorfont_external>`, the font files must be declared too in a ``.externresources.list`` file.
```

gbalan-microej replied:
Sorry, I'd forgotten about this suggestion. Updated.

rlouedec-microej commented on VEEPortingGuide/vgCco.rst:
```diff
@@ -132,9 +134,9 @@ All fonts do not require the same heap size. FreeType heap usage can be monitore
+#. The Application loads a font with `ej.microvg.VectorFont.loadFont()`_.
 
-   * If the resource is external, it is opened.
+   * If the resource is internal, it is opened.
```
Why ?

AFAIK, if the resource is internal, there is nothing to do.
But it it is external (for example, a file from some FFS), it must be opened first.
Am I wrong ?

gbalan-microej replied:
I have the impression that there's something missing for internal fonts, which is why I thought there was a problem: the first line for internal, the second for external... actually not but according to me, it misses something

gbalan-microej replied:
I understand the points & bullets. The current solution is okay for me. 

But it is not the "issue": when I read your suggestion, I have the impression that bullet # 1 only concerns external resources (even with the addition between brackets). Here's my proposal:

* If the resource is internal (or if it is an external resource located in byte-addressable memory), Freetype library simply retains the resource address (no loading step).
* If the external resource [...]

rlouedec-microej replied:
I see, it is not obvious it is is an optional step.
I pushed 3 commits, the last one clarifies that.

gbalan-microej replied:
ok for me. Just one more thing: maybe you can add a link to the chapter "external resource loader"

>    * Else, if the resource is external from non-byte-addressable memory, the FreeType library is configured to use the  :ref:`external loader<section_externalresourceloader>` to read from that memory.

rlouedec-microej replied:
Done in: 5af7719
